tests/unit/test_models.py

```diff
@@ -846,6 +848,139 @@ def test_contains_container_types(container_column, engine):
+    def verify_expected(_, *, obj, column, value):
+        nonlocal num_called, called, received
+        called = True
+        num_called += 1
+        received['obj'] = obj
+        received['column'] = column
+        received['value'] = value
+
+    class ColumnDefaultModel(BaseModel):
+        id = Column(Integer, hash_key=True, default=12)
+
+    default_obj = ColumnDefaultModel()
+
+    assert called is True
+    assert num_called == 1
+    assert received["obj"] is default_obj
+    assert received["column"] is ColumnDefaultModel.id
+    assert received["value"] == 12
+
+
+def test_column_default_func():
+
+    # Set should send the new value
+    called = False
+    num_called = 0
+    received = {}
+
+    @object_modified.connect
+    def verify_expected(_, *, obj, column, value):
+        nonlocal num_called, called, received
+        called = True
+        num_called += 1
+        received['obj'] = obj
+        received['column'] = column
+        received['value'] = value
+
+    def return_number():
+        return 123
+
+    class ColumnDefaultFuncModel(BaseModel):
+        id = Column(Integer, hash_key=True, default=return_number)
+
+    default_obj = ColumnDefaultFuncModel()
+
+    assert called is True
+    assert num_called == 1
+    assert received["obj"] is default_obj
+    assert received["column"] is ColumnDefaultFuncModel.id
+    assert received["value"] == 123
+
+
+@pytest.mark.parametrize("type, value", [
+    (String, "bob"), (UUID, uuid.uuid4()),
+    (DateTime, datetime.datetime.utcnow()), (Number, 34),
+    (Integer, 123), (Timestamp, datetime.datetime.utcnow()),
+    (Binary, 'bob'.encode()), (Boolean, False),
+    (Set(String), {'bob', 'joe', 'john'}),
+    (Set(Integer), {1, 2, 3}),
+    (Set(Binary), {b'bob', b'joe', b'john'}),
+    (List(String), ['bob', 'joe', 'john']),
+    (List(Integer), [1, 2, 3]),
+    (List(Binary), [b'bob', b'joe', b'john']),
+    (Map, {"created": DateTime, "referrer": UUID, "cache": String})
+])
+def test_column_validation(type, value):
+    class ColumnValidateModel(BaseModel):
+        class Meta:
+            validate_columns = True
+        col = Column(type, hash_key=True)
+
+    obj = ColumnValidateModel(col=value)
+    assert obj.col == value
+
+    obj.col = value
+    assert obj.col == value
+
+
+@pytest.mark.parametrize("type, value", [
+    (String, 12), (String, b'test'), (String, object()),
+    (UUID, 12), (UUID, b'test'), (UUID, object()),
+    (DateTime, 12), (DateTime, b'test'), (DateTime, object()),
+    (Number, "b12"), (Number, b'test'), (Number, object()),
+    (Integer, "b12"), (Integer, b'test'), (Integer, object()),
+    (Timestamp, "b12"), (Timestamp, b'test'), (Timestamp, object()),
+    (Binary, "b12"), (Binary, 3.14), (Binary, object()),
+    (Boolean, "b12"), (Boolean, 3.14), (Boolean, object()),
+    (Set(String), "b12"), (Set(Integer), 3.14), (Set(Binary), object()),
+    (List(String), "b12"), (List(Integer), 3.14), (List(Binary), object()),
```

Comment:
Maybe some lists that are partially-valid?  `(List(String), ["b12", 13])`
Same for other vector types.